### PR TITLE
deterministic-uname: overridable platform

### PR DESCRIPTION
**Pull request by szlend**
## Description of changes

Some packages rely on `uname` to configure the host target which breaks cross-compilation. We can have more control over the evaluation of `uname` by placing `deterministic-uname` into the package's `nativeBuildInputs`. However the current `deterministic-uname` is hardcoded to `buildPlatform`.

This PR introduces a build argument `forPlatform` to `deterministic-uname` which allows you to override the platform it reports.

Example:
```nix
deterministic-uname.override { forPlatform = stdenv.hostPlatform; }
```

### Some other ideas:

`deterministic-uname` could be based on `stdenv.hostPlatform`. Say we have:
- `system = "aarch64-darwin"`
- `crossSystem = "aarch64-linux"`
 
We could use it like so:
- `$(${pkgsBuildHost.deterministic-uname}/bin/uname) == "Linux"`
- `$(${pkgsBuildBuild.deterministic-uname}/bin/uname) == "Darwin"`.

And placing `deterministic-uname` in `nativeBuildInputs` would therefore solve cross-compilation ootb.

However this would be a backwards incompatible change. If we want to avoid that we could introduce a new variant like so:
```
host-deterministic-uname = deterministic-uname.override { forPlatform = stdenv.hostPlatform; }
```

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/applications/version-management/git/default.nix [hunk 1]
```diff
@@ -77,7 +77,9 @@ stdenv.mkDerivation (finalAttrs: {
 
   nativeBuildInputs = [ gettext perlPackages.perl makeWrapper pkg-config ]
     ++ lib.optionals withManual [ asciidoc texinfo xmlto docbook2x
-         docbook_xsl docbook_xml_dtd_45 libxslt ];
+         docbook_xsl docbook_xml_dtd_45 libxslt ]
+    ++ lib.optional (stdenv.buildPlatform != stdenv.hostPlatform) (
+         deterministic-uname.override { forPlatform = stdenv.hostPlatform; });
```
Comment (szlend) [inline pkgs/applications/version-management/git/default.nix:82]: This is an example of a package that can be fixed with a host-based `deterministic-uname`.

TODO: Remove this from the PR.

**Review by Artturin — APPROVED** [review]

**Comment by szlend** [comment]
@Artturin I moved this out of Draft and included a `deterministic-host-uname` alias for convenience. I can revert if the alias is unwanted though.

**Review by Atemu — APPROVED** [review]
The alias sounds useful, I see no reason against it.

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -3692,6 +3692,8 @@ with pkgs;
 
   deterministic-uname = callPackage ../build-support/deterministic-uname { };
 
+  deterministic-host-uname = deterministic-uname.override { forPlatform = stdenv.hostPlatform; };
```
Comment (Artturin) [inline pkgs/top-level/all-packages.nix:3695]: When this is added to `nativeBuildInputs` it'll just be the `buildPlatform` uname so it's kinda useless
```
$ nix run ".#pkgsCross.aarch64-multiplatform.buildPackages.deterministic-host-uname" -- -a
Linux nixpkgs unknown #1-NixOS Tue Jan 1 00:00:00 UTC 1980 x86_64 GNU/Linux
```

Comment (szlend) [reply pkgs/top-level/all-packages.nix:3695]: You're right, when I tested it the inputs weren't spliced :/ 

Though `nativeBuildInputs` is `pkgsBuildHost` so it should be aware of the host system that you're building for. Is it just an incorrect offset?

Comment (szlend) [reply pkgs/top-level/all-packages.nix:3695]: Switching it to `targetPlatform` solves the issue. This should be fine given everything is offset by 1 right?

Comment (szlend) [reply pkgs/top-level/all-packages.nix:3695]: Fixed in #341256 341256

**Pull request merged**